### ISSUE TITLE
Raster styling Singleband pseudocolor not working

### ISSUE THREAD

**Issue by skualos** [qgis/QGIS#37547]
Rater visualization in Singleband pseudocolor is not updated when changing trough the Min / Max options.
Steps to reproduce:
1. Load a single band raster
2. Change styling to Singleband pseudocolor
3. Choose a color ramp (otherwise no image is shown - not sure this was the case before)
4. Toggling between the Min / Max Value Settings has no effect. Changing cumulative count percentage values has no effect.
5. Select "User defined" and manually change min and max is the only one that works.

Can anyone reproduce this ?
QGIS 3.14.0 on Arch Linux

**Comment by domlysz** [comment]
You must push the classify button and it sound perfectly logic to me, IMHO editing manually min/max or selecting a new color ramp should not trigger an automatic reclassification. It's also the case when changing the interpolation method, and this last one is very tedious especially when you have worked for hours for the perfect classification and colors panel and just want to test what it will look like with another interpolation method.

There is a now closed issue about this topic : https://github.com/qgis/QGIS/issues/25001
For me it's not fixed, it's often a pain to work with singleband pseudocolor because of these abrupt auto-reclass and I feel like there is an important ui design issue behind this.

**Comment by skualos** [comment]
I guess this can make sense, but it depends on usage or personal preference IMO.
But the ability to scroll the percentage values while on the "Cumulative count cut" option, and having the classification updated automatically, is very useful behavior that I don't think conflicts with what you said.
Ideally there could be a "Automatic update" checkbox on the renderer to suit everyone's preference.

**Comment by gioman** [comment]
@domlysz @skualos 

If you open the "layer styling" panel you have the "live update" option/checkbox that does exactly what the name suggest, it updates the map symbology on the fly while the user changes symbology parameters.

From what I case see on QGIS master working with singleband grey and "live update" checked, all the changes made to the min/max values (min/max, user defined, cumulative count cut, standard deviation) are immediately reflected on the map. On the other hand using a singleband pseudo-color only setting "user defined" min/max values triggers immediate changes, all other options (including returning to "min/max") need the "classify" button to be pressed, which to me is a bit inconsistent with the nature of the "live update" checkbox. 

Opinions?

**Comment by SrNetoChan** [comment]
I agree that this is probably more a usability consistency problem. We need also to look at how the similar classification works on the vector side.

**Comment by gioman** [comment]
> I agree that this is probably more a usability consistency problem. We need also to look at how the similar classification works on the vector side.

@SrNetoChan In vector symbology, for example the graduated renderer, changes in the range values, symbol, mode all triggers immediate map changes if the "live update" check is selected.

**Comment by domlysz** [comment]
I don't think the issue has something to do with the live update, it's perfectly fine that any change in the classification is immediatley reflected in the map (updating a color, or a value or adding/removing a class). The problem is to clearly define what event should trigger an auto-reclass.

The gif below illustrate the issue with the interpolation option. As you can see all classes and colors are messed up

![ex1](https://user-images.githubusercontent.com/5639024/86476142-428b3080-bd46-11ea-9dd7-fcdecda2a7a6.gif)

On another hand switching a color ramp work as expected :

![ex4](https://user-images.githubusercontent.com/5639024/86476313-8f6f0700-bd46-11ea-9c3e-d2f73890a986.gif)

Otherwise there is a bunch of options which are related to how the automatic reclassification should perform : min/max, mode and number of classes. These options should be regrouped in a sub panel to clearly indicate that they are related to the automatic reclassification function.

Actually editing some of these option trigger an auto-reclass, for example min/max :
![ex2](https://user-images.githubusercontent.com/5639024/86476735-61d68d80-bd47-11ea-96ee-a73684caf304.gif)

Why not, but what's the point of the classify button in this case ?

If the user push the classify button, then the color ramp is upated and if the live update option is checked then the map is updated. It's a 2 step workflow, we can bypass the classify button if we want, anyway the live update only react to a change in the classification and it's fine.

**Comment by skualos** [comment]
@gioman it is also the case when working with Multiband color. Changes in Min/max are immediately applied. Singleband pseudo-color was also like this for Singleband pseudo-color before the last update. So if this wasn't on purpose, and it's not consistent with the other ones, I guess it can be considered a bug ?

**Comment by SanttuVP** [comment]
Bumped up to this issue, when I falsely thought that my singleband pseudocolor capability was not working, and I was wondering if there were any ongoing issues on the matter. As I understood after just clicking around in QGIS, you have to remember to click the classify-button or sometimes choose the color ramp before you'll get the raster styling rendered. 

Just wanted to ask whether somebody else has had this same experience and whether it would be useful to draft this idea as a feature request. Attached a few videos below to elaborate on my user experience. Sorry, if there are any problems while bringing the idea up on an issue that is not directly related to enhancing the issue in question.      

When opening the raster for the first time (no color ramp for default):

https://user-images.githubusercontent.com/13584679/122551757-18597e00-d03e-11eb-9e6f-82aa6d19a800.mp4

After using the raster before, QGIS shows a color ramp:

https://user-images.githubusercontent.com/13584679/122551781-1f808c00-d03e-11eb-94cc-f004f477cbb6.mp4

I guess my question is: **should QGIS render the raster with some default configs (color ramp, classification method) also with the singleband pseudocolor -method**, and would you consider that it would be worthwhile to draft a feature request?

**Comment by elpaso** [comment]
I believe that this is a bug: switching to singleband pseudocolor should automatically classify the raster based on the first color ramp (perhaps remembering the last color ramp used in the past instead of picking the first one in the list).

From an UX point of view, a change of renderer should always produce a valid default classification, there is no point in showing a blank raster.

**Comment by RichardScottOZ** [comment]
I have upgraded to 3.18 today

Singleband pseudocolor does go bank when changing to it.  How it used to work before, so definitely a bug.

**Comment by Pedro-Murteira** [comment]
I've tested this issue on a new profile on QGIS 3.22.3. 
1. On my first test, after selecting "Singleband pseudocolor", the raster disappears, but after choosing a Color ramp I get the color I chose. 
2. On a second test, after closing QGIS and starting a new project, if I go to "Singleband pseudocolor", QGIS remembers the last color ramp used. 

@elpaso Based on your comment, this issue seems to be working as expected on QGIS 3.22.3?

**Comment by skualos** [comment]
@Pedro-Murteira I confirm this also works as expected for me.

However the OP was about changes in the min/max values, which would only be applied after clicking the 'Classify' button.
This was a change in behaviour at that time, but now is again working the back the same way it was originally (raster is automatic restyled upon changing min/max values)

However it seems this post brought up a lot of comments on how the Singleband pseudocolor styling interface should / could work. So not sure if I should close it ?

**Comment by gioman** [comment]
> So not sure if I should close it ?

@skualos if the original report/issue has been fixed (and it seems so) this should be closed. Other issues or improvement requests can be filed in separate tickets.
